engine/src/main/java/org/terasology/engine/physics/bullet/BulletPhysics.java

```diff
@@ -128,27 +161,227 @@ public BulletPhysics() {
+                bt.setLinearFactor(temp.set(rigidBody.linearFactor));
+                bt.setFriction(rigidBody.friction);
+
+                bt.setAngularVelocity(rigidBody.angularVelocity);
+                bt.setLinearVelocity(rigidBody.velocity);
+                bt.userData = e;
+
+                bt.setWorldTransform(
+                        new Matrix4f()
+                                .translationRotateScale(
+                                        location.getWorldPosition(new Vector3f()),
+                                        location.getWorldRotation(new Quaternionf()),
+                                        1.0f));
+
+                discreteDynamicsWorld.addRigidBody(bt,
+                        CollisionGroup.combineGroups(rigidBody.collisionGroup),
+                        CollisionGroup.combineGroups(rigidBody.collidesWith));
+                return bt;
+            } else if (shape != null) {
+                shape.dispose();
+            }
+            return null;
+        });
     }
 
-    public btDiscreteDynamicsWorld getWorld() {
-        return discreteDynamicsWorld;
+
+
+    @ReceiveEvent(components = {RigidBodyComponent.class, LocationComponent.class})
+    public void updateRigidBody(OnChangedComponent event, EntityRef entity, RigidBodyComponent body, LocationComponent location) {
+        btRigidBody rigidBody = entityRigidBodies.get(entity);
+        if (rigidBody != null) {
+            if (body.kinematic) {
+                rigidBody.setCollisionFlags(rigidBody.getCollisionFlags() | btCollisionObject.CollisionFlags.CF_KINEMATIC_OBJECT);
+                rigidBody.setActivationState(Collision.DISABLE_DEACTIVATION);
+            } else {
+                rigidBody.setCollisionFlags(rigidBody.getCollisionFlags() & ~btCollisionObject.CollisionFlags.CF_KINEMATIC_OBJECT);
+                rigidBody.setActivationState(Collision.ACTIVE_TAG);
+            }
+            rigidBody.setAngularFactor(rigidBody.getAngularFactor());
+            rigidBody.setRestitution(rigidBody.getRestitution());
+            rigidBody.setLinearFactor(rigidBody.getLinearFactor());
+            rigidBody.setFriction(rigidBody.getFriction());
+
+            discreteDynamicsWorld.removeRigidBody(rigidBody);
+            discreteDynamicsWorld.addRigidBody(rigidBody,
+                    CollisionGroup.combineGroups(body.collisionGroup),
+                    CollisionGroup.combineGroups(body.collidesWith));
+        } else {
+            newRigidBody(null, entity, body, location);
+        }
     }
-    //*****************Physics Interface methods******************\\
 
-    @Override
-    public List<PhysicsSystem.CollisionPair> getCollisionPairs() {
-        List<PhysicsSystem.CollisionPair> temp = collisions;
-        collisions = new ArrayList<>();
-        return temp;
+    @ReceiveEvent(components = {RigidBodyComponent.class, LocationComponent.class})
+    public void removeRigidBody(BeforeDeactivateComponent event, EntityRef entity) {
```

Comment:
Why is it important that the entity that's deactivated has the location component?